File: test-compare.py
"""
Compare Python output with Node.js
"""
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
import binascii
from crccheck.crc import CrcArc

# Test values
client_id = '555555'
signal_type = 'BA'
zone = '01'
latitude = 59.9139
longitude = 10.7522
key = '4B38665033516D3741325A7839524465'
key_bytes = bytes.fromhex(key)

# Fixed timestamp
time_part = '12:00:00'
date_part = '01-18-2026'

alarm_command = f'Nri/{signal_type}{zone}'

print('=== Test Parameters ===')
print(f'ClientID: {client_id}')
print(f'Alarm Command: {alarm_command}')
print(f'Latitude: {latitude}')
print(f'Longitude: {longitude}')
print('')

# Python's send_alarm receives (longitude, latitude) but main() passes (lat, lon)
# So we simulate: send_alarm(..., lat_as_longitude, lon_as_latitude, ...)
def convert_coordinates(latitude, longitude):
    lon_value = float(longitude)
    lat_value = float(latitude)
    
    lon_degrees = int(lon_value)
    lon_minutes = (lon_value - lon_degrees) * 60
    lon_format = f"[X{lon_degrees:03d}E{lon_minutes:.8f}]"
    
    lat_degrees = int(lat_value)
    lat_minutes = (lat_value - lat_degrees) * 60
    lat_format = f"[Y{lat_degrees:02d}N{lat_minutes:.8f}]"
    
    return lon_format + lat_format

# In real Python code: send_alarm is called with (lat, lon) in (longitude, latitude) params
# Then convert_coordinates(latitude, longitude) swaps them back
# So we simulate: convert_coordinates(lon, lat) to match the actual behavior
output_coordinates = convert_coordinates(longitude, latitude)
print(f'=== Python Behavior ===')
print(f'Coordinates: {output_coordinates}')

input_string = f'"*SIA-DCS"0005L0#{client_id}[{client_id}|{alarm_command}]{output_coordinates}_{time_part},{date_part}'
print(f'Input String: {input_string}')

# Split at '['
part_before, part_after = input_string.split('[', 1)
part_before = f"{part_before}["
part_after = "|" + part_after
print(f'Part Before: {part_before}')
print(f'Part After: {part_after}')

# Padding
def add_padding(input_string):
    input_bytes = input_string.encode()
    pad_length = 16 - len(input_bytes) % 16
    pad_length = pad_length if pad_length != 16 else 0
    padding_ascii = bytes([i % 26 + 65 for i in range(pad_length)])
    padded_bytes = padding_ascii + input_bytes
    return padded_bytes

padded_data = add_padding(part_after)
print(f'Padded (ASCII): {padded_data.decode("ascii")}')
print(f'Padded (Hex): {padded_data.hex()}')

# Encrypt
iv = b'\x00' * 16
cipher = Cipher(algorithms.AES(key_bytes), modes.CBC(iv), backend=default_backend())
encryptor = cipher.encryptor()
encrypted_data = encryptor.update(padded_data) + encryptor.finalize()
encrypted_hex = binascii.hexlify(encrypted_data).decode('ascii').upper()
print(f'Encrypted Hex: {encrypted_hex}')

# Combine
before_and_after = f"{part_before}{encrypted_hex}"
print(f'Before+After: {before_and_after}')

# CRC and Length
crc_arc = CrcArc()
crc_arc.process(before_and_after.encode())
crc_value = crc_arc.finalhex().upper()
length_hex = f'{len(before_and_after.encode()):04X}'
print(f'CRC: {crc_value}')
print(f'Length: {length_hex}')
print(f'CRC+Length: {crc_value}{length_hex}')

# Final message
with_hash = f"{crc_value}{length_hex}{before_and_after}"
full_message = f"\r\n{with_hash}\r"
message_bytes = full_message.encode()

print('')
print('=== Final Message ===')
print(f'ASCII (escaped): {full_message.replace(chr(13), chr(92)+"r").replace(chr(10), chr(92)+"n")}')
print(f'Hex: {message_bytes.hex().upper()}')
print(f'Length: {len(message_bytes)} bytes')

File: test-compare.js
/**
 * Compare Node.js output with Python output
 */

const { 
  addPadding, 
  encryptAES, 
  calculateCRC, 
  convertCoordinates 
} = require('./src/sia-utils');

// Test values - same as we'd use in Python
const clientId = '555555';
const signalType = 'BA';
const zone = '01';
const latitude = 59.9139;  // User enters as Latitude
const longitude = 10.7522;  // User enters as Longitude
const key = '4B38665033516D3741325A7839524465';

// Fixed timestamp for comparison
const timePart = '12:00:00';
const datePart = '01-18-2026';

const alarmCommand = `Nri/${signalType}${zone}`;

console.log('=== Test Parameters ===');
console.log('ClientID:', clientId);
console.log('Alarm Command:', alarmCommand);
console.log('Latitude:', latitude);
console.log('Longitude:', longitude);
console.log('');

// Python's convert_coordinates receives (latitude, longitude) from send_alarm
// But send_alarm's params are named (longitude, latitude) - they're swapped in main()!
// main() calls: send_alarm(..., float(PosistionLat), float(PositionLong), ...)
// send_alarm signature: send_alarm(alarm_command, user_id, longitude, latitude, ...)
// So: PosistionLat goes to "longitude" param, PositionLong goes to "latitude" param
// Then convert_coordinates(latitude, longitude) uses the swapped values

// In Python's convert_coordinates:
// - longitude param (which is actually user's PositionLong=10.7522) -> X format
// - latitude param (which is actually user's PosistionLat=59.9139) -> Y format

// So the actual Python logic for coordinates:
// X format uses: what user entered as Longitude (10.7522)
// Y format uses: what user entered as Latitude (59.9139)

// Let's trace Python more carefully:
// main(): PosistionLat=59.9139, PositionLong=10.7522
// main() calls: send_alarm(..., float(PosistionLat), float(PositionLong), ...)
//                              = send_alarm(..., 59.9139, 10.7522, ...)
// send_alarm signature: send_alarm(alarm_command, user_id, longitude, latitude, ...)
// So: longitude=59.9139, latitude=10.7522
// Then: convert_coordinates(latitude, longitude) = convert_coordinates(10.7522, 59.9139)
// In convert_coordinates(latitude, longitude):
//   lon_value = longitude = 59.9139
//   lat_value = latitude = 10.7522
// So X format uses 59.9139 (labeled as lon but is actually lat)
// And Y format uses 10.7522 (labeled as lat but is actually lon)

// This matches Python's buggy parameter naming
// Python passes (lat, lon) to (longitude, latitude) params in send_alarm
// Then convert_coordinates(latitude, longitude) makes X use the "longitude" (which is lat) 
// and Y use the "latitude" (which is lon)
// So we call convertCoordinates(longitude, latitude) to match this behavior
console.log('=== Python Behavior (Matching the swap) ===');
const pythonCoords = convertCoordinates(longitude, latitude);  // Match Python: (10.7522, 59.9139)
console.log('Python-style Coordinates:', pythonCoords);

// Build input_string like Python
const inputString = `"*SIA-DCS"0005L0#${clientId}[${clientId}|${alarmCommand}]${pythonCoords}_${timePart},${datePart}`;
console.log('Input String:', inputString);

// Split at '['
const bracketIndex = inputString.indexOf('[');
const partBefore = inputString.substring(0, bracketIndex + 1);
const partAfter = '|' + inputString.substring(bracketIndex + 1);
console.log('Part Before:', partBefore);
console.log('Part After:', partAfter);

// Padding
const paddedData = addPadding(Buffer.from(partAfter, 'ascii'));
console.log('Padded (ASCII):', paddedData.toString('ascii'));
console.log('Padded (Hex):', paddedData.toString('hex'));

// Encrypt
const encryptedHex = encryptAES(paddedData, key);
console.log('Encrypted Hex:', encryptedHex);

// Combine
const beforeAndAfter = `${partBefore}${encryptedHex}`;
console.log('Before+After:', beforeAndAfter);

// CRC and Length
const crcArc = calculateCRC(beforeAndAfter);
const lengthHex = beforeAndAfter.length.toString(16).toUpperCase().padStart(4, '0');
console.log('CRC:', crcArc);
console.log('Length:', lengthHex);
console.log('CRC+Length:', crcArc + lengthHex);

// Final message
const withHash = `${crcArc}${lengthHex}${beforeAndAfter}`;
const fullMessage = `\r\n${withHash}\r`;
console.log('');
console.log('=== Final Message ===');
console.log('ASCII (escaped):', fullMessage.replace(/\r/g, '\\r').replace(/\n/g, '\\n'));
console.log('Hex:', Buffer.from(fullMessage, 'ascii').toString('hex').toUpperCase());
console.log('Length:', fullMessage.length, 'bytes');
